main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+
+      case MonoAst.Expr.NewObject(name, clazz, tpe, eff, methods, loc) =>
+        val (ms, o) = visitJvmMethods(methods)
+        (OccurrenceAst1.Expr.NewObject(name, clazz, tpe, eff, ms, loc), increment(o))
+
+    }
+
+    /**
+      * Performs occurrence analysis on a list of expressions `exps` and merges occurrences.
+      * Captures `sym0`.
+      */
+    def visitExps(exps: List[MonoAst.Expr])(implicit letBinding: Option[VarSym]): (List[OccurrenceAst1.Expr], OccurInfo) = {
+      val (es, o1) = exps.map(visit).unzip
+      val o2 = o1.foldLeft(OccurInfo.Empty)(combineInfo)
+      (es, o2)
+    }
+
+    def combineApplyCloInfo(occurInfo0: OccurInfo, exp0: OccurrenceAst1.Expr): OccurInfo = exp0 match {
+      case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.Closure(sym), _, _, _, _) =>
+        occurInfo0 :+ sym -> Once
+      case _ => occurInfo0
+    }
+
+    def visitAtomicOp(op0: AtomicOp, occurInfo0: OccurInfo)(implicit letBinding: Option[VarSym]): OccurInfo = op0 match {
+      case AtomicOp.Is(sym) if sym.name == "Choice" =>
+        occurInfo0 :+ sym0 -> DontInline
+
+      case AtomicOp.HoleError(_) => letBinding match {
+        case Some(varSym) => occurInfo0 + (varSym -> DontInline)
+        case None => occurInfo0
+      }
+
+      case AtomicOp.MatchError => letBinding match {
+        case Some(varSym) => occurInfo0 + (varSym -> DontInline)
+        case None => occurInfo0
+      }
+
+      case AtomicOp.Cast => occurInfo0 :+ sym0 -> Dangerous
+
+      case _ => occurInfo0
+    }
+
+    def visitMatchRules(rules0: List[MonoAst.MatchRule])(implicit letBinding: Option[VarSym]): (List[OccurrenceAst1.MatchRule], OccurInfo) = {
+      val (rs, o) = rules0.map {
+        case MonoAst.MatchRule(pat, guard, exp) =>
+          val (g, o1) = guard.map(visit).unzip
+          val (e, o2) = visit(exp)
+          val o3 = combineInfoOpt(o1, o2)
+          val (p, syms) = visitPattern(pat, o3)
+          val o4 = o3 -- syms
+          (OccurrenceAst1.MatchRule(p, g, e), o4)
+      }.unzip
+      val o1 = o.foldLeft(OccurInfo.Empty)(combineInfoBranch)
+      (rs, o1)
+    }
+
+    def visitTryCatchRules(rules0: List[MonoAst.CatchRule])(implicit letBinding: Option[VarSym]): (List[OccurrenceAst1.CatchRule], OccurInfo) = {
+      val (rs, o) = rules0.map {
+        case MonoAst.CatchRule(sym, clazz, exp) =>
+          val (e, o) = visit(exp)
+          (OccurrenceAst1.CatchRule(sym, clazz, e), o)
+      }.unzip
+      val o1 = o.foldLeft(OccurInfo.Empty)(combineInfo)
+      (rs, o1)
+    }
+
+    def visitTryWithRules(rules0: List[MonoAst.HandlerRule])(implicit letBinding: Option[VarSym]): (List[OccurrenceAst1.HandlerRule], OccurInfo) = {
+      val (rs, o) = rules0.map {
+        case MonoAst.HandlerRule(op, fparams, exp) =>
+          val (e, o) = visit(exp)
+          val fps = fparams.map(visitFormalParam)
+          (OccurrenceAst1.HandlerRule(op, fps, e), o)
+      }.unzip
+      val o1 = o.foldLeft(OccurInfo.Empty)(combineInfo)
```

Comment:
also branch maybe? although handlers are a bit hard to think about. They are almost like a local def